docs: Design document for Block Verification Process (#384)
hashgraph/hedera-block-node: server/docs/design/block-verification.md

Reviewer: General note, at a bare minimum, we should wrap lines to no more than 120 columns, the same as the java files.

Author: ✅ Wrapped the whole document to `80`

  - Accepts block items incrementally. (continues to compute them asynchronously)
  - Once the block_proof is provided, finalizes the hash computation asynchronously.
  - After computing the final hash, calls SignatureVerifier for verification.
- ### SignatureVerifier
  - Verifies the signature by comparing the computed hash to the hash implied by the signature (using the public key).
  - Report results to BlockStatusManager.
- ### BlockStatusManager
  - Receives verification results from SignatureVerifier.
  - Updates block status and triggers any necessary recovery or follow-up processes depending on the outcome.

## Design
1. The `VerificationHandler` receives the list of block items from the unverified ring buffer.
1. When the block_header is detected, the `VerificationHandler` creates a `BlockHashingSession` using the `BlockHashingSessionFactory`.
1. The `BlockHashingSession` accepts subsequent block items incrementally.
1. Once the block_proof is received, the `BlockHashingSession` calls `completeHashing()` to finalize the hash computation.
1. Upon completion of computing the final block hash, the `BlockHashingSession` calls the `SignatureVerifier` to verify the signature.

> 1. The `SignatureVerifier` compares the computed hash to the hash implied by the signature using the public key.

Reviewer: ```suggestion
1. The `SignatureVerifier` compares the computed hash to the hash signed by the Block Proof signature.
```

Author: ✅ Done!

1. [Interfaces](#interfaces)
1. [Enums](#enums)
1. [Metrics](#metrics)
1. [Exceptions](#exceptions)

## Purpose
The purpose of the Block Verification feature is to ensure that blocks received from consensus nodes are valid and have not been tampered with. This is achieved by re-calculating the block hash and verifying it against the signature provided by the consensus node.

## Goals
1. The block-node must re-create the block hash from the block items and verify that it matches the hash implied by the signature. 
1. If verification fails, the block should be considered invalid, and appropriate error-handling procedures must be triggered.

## Terms
<dl>
<dt>Consensus Node (CN)</dt><dd>A node that produces and provides blocks.</dd>
<dt>Block Items</dt><dd>The block data pieces (header, events, transactions, transaction result, state changes, proof) that make up a block.</dd>
<dt>Block Hash</dt><dd>A cryptographic hash representing the block’s integrity.</dd>
<dt>Signature</dt><dd>The cryptographic signature of the block hash created by Network private key (aka: LedgerId).</dd>
<dt>Public Key</dt><dd>The public key of the LedgerId that signed the block.</dd>
</dl>

> ## Entities

Author: I would push back on this one, since Entities have a bullet list and more than 1 line of explanation, not sure on documentation guidelines and best practices, but does not seem right to me to make Entities a definition list.

> ## Entities

Reviewer: I would say it can go either way.
Whichever approach produces the most readable layout is fine, and I believe we can trust one another (with occasional suggestions for possible alternatives) to choose the best option.

> ## Entities

Reviewer 2: @jsync-swirlds +1, @AlfredoG87 up to you.

        
    U->>V: (1) onBlockItemsReceived(List<BlockItem>)        
    
    alt (2) Detects block_header
    
    V->>F: createSession(initialBlockItems, executorService, signatureVerifier)
    Note over S: New instance of BlockHashingSession created
    F-->>V: returns BlockHashingSession (S)
    V-->>U: Returns without blocking
    S->>S: *Starts hash computation asynchronously
    
    else (3) Append more Block Items
    loop
    V->>S: addBlockItems(items)
    V-->>U: return without blocking
    S->>S: *Continues hash computation asynchronously
    end

    else (4) Append BlockItems with block_proof
   
    V->>S: addBlockItems(items with block_proof)
    V-->>U: return without blocking    
    S->>S: async completeHashing()   
    S->>SV: (5) verifySignature(signature, computedHash, blockNumber)    

    Note over SV,BSM: (6) Compare computed hash and signature
    alt (7) Verification Fails
      SV->>BSM: updateBlockStatus(blockNumber, ERROR, SIGNATURE_INVALID      
    else (8) Verification Succeeds
      SV->>BSM: updateBlockStatus(blockNumber, VERIFIED, NONE)      
    end
    Note over BSM: (9) Follow-up to downstream services
    end

```

> ## Interfaces

Author: I guess that the intention is to complement the diagram and design overview, documentation should be kept up to date regardless and interfaces should not change that often. I belive they add value, but if the general consensus is to remove them, I don't mind either. Am good either way 👍

> ## Interfaces

Reviewer: It's reasonable to include some interface definitions, but it's also important to limit detail. I would suggest that good judgement be exercised to keep the detail at an appropriate level; I trust @AlfredoG87 can achieve the right balance in this section.

### VerificationHandler
```java
public interface VerificationHandler {
  void onBlockItemsReceived(List<BlockItem> blockItems);  
}
```

### BlockHashingSessionFactory
```java
import java.util.concurrent.ExecutorService;

public interface BlockHashingSessionFactory {
  BlockHashingSession createSession(List<BlockItem> initialBlockItems, ExecutorService executorService, SignatureVerifier signatureVerifier);
}
```

### BlockHashingSession
```java
public interface BlockHashingSession {
  void addBlockItem(BlockItem item);
  CompletableFuture<Void> completeHashing();
}
```

### SignatureVerifier
```java
public interface SignatureVerifier {
  void verifySignature(byte[] signature, byte[] computedHash, long blockNumber);  
}
```

### BlockStatusManager
```java

public interface BlockStatusManager {
    void updateBlockStatus(long blockNumber, BlockVerificationStatus status, VerificationError error);
}
```

## Enums
```java
public enum BlockVerificationStatus {
  VERIFIED,
  ERROR
}

public enum VerificationError {
  NONE,
  SIGNATURE_INVALID,
  SYSTEM_ERROR
}
```

> ## Configuration

Author: I agree that we should not duplicate documentation, as in the case of keeping configuration here and in the Config document.

I added this section since we need to highlight that we will need a `VerificationConfig` Object and that we need to include at least a way to enable or disable the verification handler, agreed that the final implementation will change how to do it and even the name and possible values for the configurations. so we want to just note that we will take into consideration some but not limited to, configuration properties.

✅ Done!

## Metrics
<dl>
<dt>blocks_received</dt><dd>Counter of the number of blocks received for verification.</dd>
<dt>blocks_verified</dt><dd>Counter of the number of blocks verified.</dd>
<dt>blocks_unverified</dt><dd>Counter of the number of blocks that have not been verified, is the difference between blocks_received and blocks_verified.</dd>
<dt>blocks_signature_invalid</dt><dd>Counter of the number of blocks with invalid signatures.</dd>
<dt>blocks_system_error</dt><dd>Counter of the number of blocks with system errors.</dd>
<dt>block_verification_time</dt><dd>Histogram of the time taken to verify a block, gives the node operator an idea of the time taken to verify a block.</dd>
</dl>

## Exceptions
- **SYSTEM_ERROR:** Issues with node configuration or bugs. The node logs details, updates metrics, and might attempt recovery or halt.
- **SIGNATURE_INVALID:** If verification fails, SIGNATURE_INVALID is used. The block is marked invalid, and the BlockStatusManager triggers error-handling routines (requesting re-sends, removing corrupted data, notifying subscribers, etc.).

### Signature invalid

> If the computed hash does not match the hash signed by the network, the block is considered unverified. It is marked as such and publishers are requested to resend the block.

Reviewer: Github hides the trailing newline.  If it's actually missing there will be a red !["no-entry" "octicon" symbol](https://raw.githubusercontent.com/primer/octicons/refs/heads/main/icons/no-entry-24.svg) at the end of the file.

# Block Verification Design

## Table of Contents
1. [Purpose](#purpose)
1. [Goals](#goals)
1. [Terms](#terms)
1. [Entities](#entities)
1. [Design](#design)
1. [Interfaces](#interfaces)
1. [Enums](#enums)
1. [Metrics](#metrics)
1. [Exceptions](#exceptions)

## Purpose
The purpose of the Block Verification feature is to ensure that blocks received from consensus nodes are valid and have not been tampered with. This is achieved by re-calculating the block hash and verifying it against the signature provided by the consensus node.

## Goals
1. The block-node must re-create the block hash from the block items and verify that it matches the hash implied by the signature. 
1. If verification fails, the block should be considered invalid, and appropriate error-handling procedures must be triggered.

## Terms
- Consensus Node (CN): A node that produces and provides blocks.
- Block Items: The block data pieces (header, events, transactions, transaction result, state changes, proof) that make up a block.
- Block Hash: A cryptographic hash representing the block’s integrity.
- Signature: The cryptographic signature of the block hash created by Network private key (aka: LedgerId)

> - Public Key: The public key of the LedgerId that signed the block.

Reviewer: This is another place where a definition list is probably clearer than the bullet list.

Author: Indeed it makes sense, it should be done! 

✅ Done!

# Block Verification Design

## Table of Contents

1. [Purpose](#purpose)
2. [Goals](#goals)
3. [Terms](#terms)
4. [Entities](#entities)
5. [Design](#design)
6. [Interfaces](#interfaces)
7. [Enums](#enums)
8. [Metrics](#metrics)
9. [Exceptions](#exceptions)

## Purpose

The purpose of the Block Verification feature is to ensure that blocks received from consensus nodes are valid and have not been tampered with. This is achieved by re-calculating the block hash and verifying it against the signature provided by the consensus node.

## Goals

> 1) The block-node must re-create the block hash from the block items and verify that it matches the hash implied by the signature.
> 2) If verification fails, the block should be considered invalid, and appropriate error-handling procedures must be triggered.

Reviewer: Markdown standard is a `.` for ordered lists
```suggestion
1. The block-node must re-create the block hash from the block items and verify that it matches the hash implied by the signature.
1. If verification fails, the block should be considered invalid, and appropriate error-handling procedures must be triggered.
```

Author: ✅ Done!

## Terms

- Consensus Node (CN): A node that produces and provides blocks.
- Block Items: The block data pieces (header, events, transactions, transaction result, state changes, proof) that make up a block.
- Block Hash: A cryptographic hash representing the block’s integrity.
- Signature: The cryptographic signature of the block hash created by Network private key (aka: LedgerId)
- Public Key: The public key of the LedgerId that signed the block.

## Entities

> - **VerificationHandler** 

Reviewer: These "header" items might read better as headings.
```suggestion
- #### VerificationHandler
```

Author: ✅ Done!

> - **VerificationHandler** 

Author: We already have an interface for the handlers, is the same that the Persistence and other subscribers are using, from the unparsed and unverified ring buffer of block items.
Currently we are relying in configuration for deciding which features we enable, we even have NoOp implementations that are effectively disabling features of the system. I agree that using a Loader Subsystem could make it more extensible in the long run, however for the purpose and scope of this Epic (Block Verification) I would prefer to keep using the same architecture as PersistenceHandler, and add a configuration via a `VerificationConfig` class that enables or disables such handler (basically another consumer of the incoming ring buffer)

> - **VerificationHandler** 

Reviewer: I completely agree that the suggested change would be a significant refactor.  I'm suggesting that we should start that discussion _separate from this PR_ and get the high level design worked out and documented so we can begin that refactor in the near future (before we design too many more services).

> - **VerificationHandler** 

Author: Have captured those ideas in this [discussion](https://github.com/hashgraph/hedera-block-node/discussions/400).

  - Receives the stream of block items in the form of List<BlockItemsUnparsed> from the unparsed and unverified block items ring buffer. 
  - When it detects a block_header, it creates a BlockHashingSession using the BlockHashingSessionFactory, providing it with the initial block items (and internally, the session will handle asynchronous hashing).
  - Adds subsequent block items to the session, including the block_proof.

>   - Once the block_proof is received, calls completeHashing() on the BlockHashingSession, then releases the thread back to the UnverifiedRingBuffer.

Author: ✅ Done!

  - Does not block waiting for verification; the hash computation and verification continue asynchronously.

> - **BlockHashingSessionFactory**

Author: I think the Builder Pattern is useful when we need to provide `optional` parameters and/or pass around the builder for appending parameters from different places of code and finally build it. In this case we will always set the same parameters and the Factory in itself is a Singleton that use the same `ExecutorService` for all the instances that the factory mints.

So for this case in particular I would defend the use of a Factory. However I have consider keeping the ExecutorService instance at the Handler level, and just remove the need of a Builder or Factory, just instance the BlockHashingSession Classes directly on the VerificationHandler.

> - **BlockHashingSessionFactory**

Reviewer: We should probably discuss use of executor services; we would probably be better off using a ForkJoinPool and have that managed at higher level (Handler, or service).
Another item where this high level design _should_ (in my opinion) be ahead of the detail design, not trailing it.

> - **BlockHashingSessionFactory**

Author: I am open to discussion on how to better achieved the `Async` work, am suggesting the ExecutorService due to how the hashing algorithm is implemented on the services codebase: see [here](https://github.com/hashgraph/hedera-services/blob/8d8a21b9d60232b7565f0e8ebb0db3e2dcfec4d5/hedera-node/hedera-app/src/main/java/com/hedera/node/app/blocks/impl/BlockStreamManagerImpl.java#L237-L238).

> - **BlockHashingSessionFactory**

Reviewer: We might describe things in a bit higher level and avoid specific implementations, then, and leave the specifics to the detail design.

I don't think we should try to use the same concepts everywhere, though, and Services has a very different threading environment than Block Node.

  - Creates new BlockHashingSession instances, provides them with a ExecutorService.
- **BlockHashingSession**

>   - Holds all necessary block data.

Author: I guess you are right! 👍 
Will make the correction.

>   - Holds all necessary block data.

Author: ✅  Done!

  - Accepts block items incrementally. (continues to compute them asynchronously)
  - Once the block_proof is provided, finalizes the hash computation asynchronously.
  - After computing the final hash, calls SignatureVerifier for verification.
- **SignatureVerifier**

>   - Verifies the signature by comparing the computed hash to the hash implied by the signature (using the public key).

Author: ✅ Done!

>   - Verifies the signature by comparing the computed hash to the hash implied by the signature (using the public key).

Reviewer: This isn't visible in the latest update...

>   - Calls BlockStatusManager with the verification result.

Author: changed the wording to "Report results to BlockStatusManager" however I think this is related with the removal of that class altogether. we can further explore and discuss this in next daily scrum parking lot avaiable.

- **BlockStatusManager**
  - Receives verification results from SignatureVerifier.

>   - Updates block status and triggers any necessary recovery or follow-up processes depending on the outcome.

Author: I guess that we could create a "Verified Ring Buffer" that the SignatureVerifier writes the block there. maybe even already parsed, so that the `StateHandler` can pick it up from there. instead of relying on a coordinator/manager.

However how things are working now, the server is returning `ack` messages to the `producer` and it knows when to return them using another manager `Notifier` right now the Notifier returns the ack to the producer when the block finishes persistence, however it should be done once the block is verified, so my short term idea for this solution is to have a BlockManager that basically holds the last few blocks in memory, and gets updates of the blocks as it goes throught the pipelines (persistence, verification, stateUpdater, etc...) the BlockManager not only keeps context of `work` that was performed on the last few blocks such as `parsing`, `verification`, but also orchestrates the workflow (journey of a block_items)

>   - Updates block status and triggers any necessary recovery or follow-up processes depending on the outcome.

Author: In this case the intent of the design doc is to communicate the idea of what we intent to do in the short and immediate future, **current Verification Epic**. and once we settle on a new design for long term we should make part of that effort to update the corresponding docs. I have captured the ideas coming off this PR review in a separate [discussion](https://github.com/hashgraph/hedera-block-node/discussions/400) that we can create an Epic from.

>   - Updates block status and triggers any necessary recovery or follow-up processes depending on the outcome.

Reviewer: Fair.  We should still be careful not to paint ourselves in a corner in this document, but that's a judgement call.

## Design

> 1. The `VerificationHandler` receives the list of block items from the unverified ring buffer.

Author: Since we are getting a list of items, that is the same interface that we are using for the rest of the services. and other services. I believe that this is related to the definition of interfaces for services/modules of the system and I think we should take it out of this scope and discuss it independent of this Epic, since we already have an interface in place.

> 1. The `VerificationHandler` receives the list of block items from the unverified ring buffer.

Author: Same, moving this to another [discussion](https://github.com/hashgraph/hedera-block-node/discussions/400) so we can tackle the communication interface refactor on a separate Epic.

> 1. The `VerificationHandler` receives the list of block items from the unverified ring buffer.

Reviewer: I don't think this is the same topic as that discussion.
Here I think we have an incorrect interface definition, rather than a more broad "how do we have independent services" discussion.

> 1. The `VerificationHandler` receives the list of block items from the unverified ring buffer.

Author: Have made the pertaining corrections on the design. 👍

> 1. The `VerificationHandler` receives the list of block items from the unverified ring buffer.

Reviewer 3: We're receiving a stream of collections of BlockItems from the CN. We're emitting a stream of collections of BlockItems to consumers. I'm confused by the proposal to loop through the inbound collection to broadcast single items on the ring buffer. This is just added looping computation when the goal is to get the data as quickly as possible to the components that require them.

> 1. The `VerificationHandler` receives the list of block items from the unverified ring buffer.

Reviewer: I believe we should not be requiring the _same_ level of batching at every step.
CN might send 100 at a time, then Verification sends 1000 at a time to hashing, and each subscriber makes a different request for batch size based on their own time/overhead tradeoffs.
Sending everything internally in the same batch as we receive removes all flexibility and control and puts it at "whatever the publisher sent" with no ability to change that.

We _may_ need to be careful how often we send data between threads (that's a tricky topic and we need to test the impact in our system, rather than blindly following what Consensus Nodes experience in their much more complex system), but I think we can be reasonable to batch things up to an appropriate level _where needed_ and by sending everything into the ring buffer individually we gain both greater simplicity (the ring buffer is always single items, regardless of external entities) and more flexibility to control any internal batch sizes.

We also get the advantage that we aren't looping through lists in _every_ service trying to detect headers and proof items; we just handle each item as it comes in and if it's a header or proof then we know a block just started or ended.

So we trade one loop on intake (where it's simple and easy, and there really isn't anything else for that component to do) for multiple loops in other services, which seems a good trade in my view.

> 1. The `VerificationHandler` receives the list of block items from the unverified ring buffer.

Author: great discussion @mattp-swirldslabs and @jsync-swirlds, however we already created a place for that.

I agree that it implicates design changes to the VerificationHandler, but it also would do so for the rest of the current architecture, so for the time being we decided on take out of the current scope, and move it to a discussion #400 where we can come to an agreement and then create an EPIC for doing the necessary refactors.

2. When the block_header is detected, the `VerificationHandler` creates a `BlockHashingSession` using the `BlockHashingSessionFactory`.

> 3. The `BlockHashingSession` accepts subsequent block items incrementally.

Reviewer: Just a note, there's no need to keep the numbers up, they can all be `1.`.  Markdown translates to an ordered list, and the renderer (browser) generates all the numbers.

Author: ✅ Done!

> 4. Once the block_proof is received, the `BlockHashingSession` calls `completeHashing()` to finalize the hash computation.

Author: I've changed to how you suggested, and is how I have it here, I guess I also thought about it and it made sense, but never updated the rest of the docs. 
Thanks for noticing.

✅ Done!

5. Upon completion of computing the final block hash, the `BlockHashingSession` calls the `SignatureVerifier` to verify the signature.
6. The `SignatureVerifier` compares the computed hash to the hash implied by the signature using the public key.

> 7. If the verification fails, the `SignatureVerifier` calls the `BlockStatusManager` to update the block status as SIGNATURE_INVALID.

Author: I also thought about this "ideas" the memory of keeping previous hashes, double check with our own `state` Merkle tree, however, in the sake of KISS decided to leverage the statement that **all blocks are self contained** and keep the verification of each individual block as stateless as possible, by design. 

Adding such double checks are a nice addition, but definitively not needed at a Phase1 implementation of the BlockVerification feature, so maybe we can create a follow-up Epic for BlockVerification Phase2 tasks and add these ideas in there.

> 7. If the verification fails, the `SignatureVerifier` calls the `BlockStatusManager` to update the block status as SIGNATURE_INVALID.

Reviewer: I agree that such enhancements should be a separate effort.
I disagree that we should not describe high level plans for additional work in this document; there's no reason. that I know of, to restrict the design work here to _only_ what will be competed in the current epic.

> 7. If the verification fails, the `SignatureVerifier` calls the `BlockStatusManager` to update the block status as SIGNATURE_INVALID.

Author: Yes, I agree that a design document can serve both as a roadmap and/or a blueprint, but at this level for this project, we have been doing the latter. Using them as a tool for socializing implementation.

> 7. If the verification fails, the `SignatureVerifier` calls the `BlockStatusManager` to update the block status as SIGNATURE_INVALID.

Reviewer: "We have been doing" is not convincing, to me.  We may need to discuss in the team, but I believe that there is little or no value to writing these documents _after_ writing the detail design (we already have full detail, what do we need high-level for at that point?).
Side note, code isn't implementation, it's just detailed design.  Implementation is what the build system produces (automatically).

My opinion (which is not controlling; the team may not agree, at which point we'll document the team decision and commit to that):
Design documents should be where we discuss and debate the high-level design approach for a proposed component before doing the detail design work documented in code files.  This would be where we determine if we have appropriate interfaces, the sequence of events is correct, the core data structures are reasonable, the algorithmic designs are effective, etc...
Documenting a consensus on that high-level design before doing the detail work is valuable enough to warrant a document.  I would contend, however, that "socializing" a design after the fact is much less valuable, and might not be worth the cost.

8. If the verification succeeds, the `SignatureVerifier` calls the `BlockStatusManager` to update the block status as VERIFIED.

> 9. The `BlockStatusManager` triggers any necessary recovery or follow-up processes depending on the verification result.

Author: ✅ Done!

Yes, I leaved a comment on the suggestion, am open to it.

Sequence Diagram:
```mermaid
sequenceDiagram
    participant U as UnverifiedRingBuffer
    participant V as VerificationHandler
    participant F as BlockHashingSessionFactory
    participant S as BlockHashingSession
    participant SV as SignaturesequenceDiagram
    participant U as UnverifiedRingBuffer
    participant V as VerificationHandler
    participant F as BlockHashingSessionFactory
    participant S as BlockHashingSession
    participant SV as SignatureVerifier
    participant BSM as BlockStatusManager

        
    U->>V: (1) onBlockItemsReceived(List<BlockItem>)        
    
    alt (2) Detects block_header
    
    V->>F: createSession(initialBlockItems, executorService, signatureVerifier)
    Note over S: New instance of BlockHashingSession created
    F-->>V: returns BlockHashingSession (S)

>     V-->>U: Returns without blocking

Reviewer: Note, minimizing time on ring buffer threads is one part of making this performance equation work.
The other is possibly changing clients to use a blocking poll rather than a callback/reactive approach.  Certainly not what we need/want to do at the moment, but it may well be worth thinking about how that would work in a future update.

>     V-->>U: Returns without blocking

Reviewer: This matches how I understood it, and our slack discussion was helpful.

However, I think we should at least consider Jasper's suggestion that we try polling and/or blocking polling for at least some of the Ring Buffer clients.  We may be able to further reduce thread transitions and avoid some of the thread-lock possibilities inherent in the "push" option.

>     rect rgb(230, 230, 230)

Author: I understand, thanks for the suggestion, changed it for an orange, however depending on what we decide, we might just remove the different background to represent future/optional implementation that might just be implemented from the start.

public interface VerificationHandler {
  void onBlockItemsReceived(List<BlockItem> blockItems);  
}
```
### BlockHashingSessionFactory

```java
import java.util.concurrent.ExecutorService;

public interface BlockHashingSessionFactory {
  BlockHashingSession createSession(List<BlockItem> initialBlockItems, ExecutorService executorService, SignatureVerifier signatureVerifier);
}
```
### BlockHashingSession
```java
/* Once hashing is completed internally, calls SignatureVerifier to continue with verification process */
public interface BlockHashingSession {
  void addBlockItem(BlockItem item);
  CompletableFuture<Void> completeHashing(); // triggers final hash computation asynchronously  
}
```
### SignatureVerifier
```java
public interface SignatureVerifier {
  void verifySignature(byte[] signature, byte[] computedHash, long blockNumber);  
}
```
### BlockStatusManager
```java

public interface BlockStatusManager {
    void updateBlockStatus(long blockNumber, BlockVerificationStatus status, VerificationError error);
}
```

## Enums
```java
public enum BlockVerificationStatus {
  VERIFIED,
  ERROR
}

public enum VerificationError {
  NONE,
  SIGNATURE_INVALID,
  SYSTEM_ERROR
}

> ```

Reviewer: I generally find any code at this level of design to be too much detail, but I suppose as long as we limit this to interface only, it may be OK.

Author: Yes, I decided to add them at the Design level to help clarify the relationships between classes. However am also wary of adding too much low level code details to a design doc, since that adds work to keep it up to date when things change on the actual code implementation. But interfaces should not change that much and is easier to maintained them so I thought that the trade-off was worth it.

## Metrics

> - **blocks_received**: Counter of the number of blocks received for verification
> - **blocks_verified**: Counter of the number of blocks verified
> - **blocks_unverified**: Counter of the number of blocks that have not been verified, is the difference between blocks_received and blocks_verified.
> - **blocks_signature_invalid**: Counter of the number of blocks with invalid signatures
> - **blocks_system_error**: Counter of the number of blocks with system errors
> - **block_verification_time**: Histogram of the time taken to verify a block, gives the node operator an idea of the time taken to verify a block.

Reviewer: Sadly, github doesn't support markdown definition lists directly, but you _could_ write this as one using tags (and the result would, in my opinion, be more readable).
```suggestion
<dl>
<dt>blocks_received</dt><dd>Counter of the number of blocks received for verification.</dd>
<dt>blocks_verified</dt><dd>Counter of the number of blocks verified.</dd>
<dt>blocks_unverified</dt><dd>Counter of the number of blocks that have not been verified, is the difference between blocks_received and blocks_verified.</dd>
<dt>blocks_signature_invalid</dt><dd>Counter of the number of blocks with invalid signatures.</dd>
<dt>blocks_system_error</dt><dd>Counter of the number of blocks with system errors.</dd>
<dt>block_verification_time</dt><dd>Histogram of the time taken to verify a block, gives the node operator an idea of the time taken to verify a block.</dd>
</dl>
```
That renders like this:
<dl>
<dt>blocks_received</dt><dd>Counter of the number of blocks received for verification.</dd>
<dt>blocks_verified</dt><dd>Counter of the number of blocks verified.</dd>
<dt>blocks_unverified</dt><dd>Counter of the number of blocks that have not been verified, is the difference between blocks_received and blocks_verified.</dd>
<dt>blocks_signature_invalid</dt><dd>Counter of the number of blocks with invalid signatures.</dd>
<dt>blocks_system_error</dt><dd>Counter of the number of blocks with system errors.</dd>
<dt>block_verification_time</dt><dd>Histogram of the time taken to verify a block, gives the node operator an idea of the time taken to verify a block.</dd>
</dl>

Author: Greate Suggestion, will try to remember for future MD hacks 💯 

✅ Done!

- 

> ## Exceptions

Reviewer: Are these exceptions or response codes?

Author: I think we can further refine this, due to the Async Nature of verification handler, exceptions becomes response codes or statuses of the work done. So I would say that in this case they are both.

Author: I suppose that this is related but out of scope for now we should capture the missing and pending work so we can take into consideration when designing the inter-module communication protocol standard.
There will be a Part 2 of Verification work, when we will actually verify the signature not just calculating the hash. At the time we will have more parts for the system defined and hopefully a inter-module communication protocol standard that defines what and how are exceptions and results handled.

Author: https://github.com/hashgraph/hedera-block-node/issues/451

Reviewer: I still recommend using cleaner language here; as this is a _design_ document, not merely a _report_.

Resolving for now, as the proper role and tone for high level design documents is outside the scope of a single PR.

- **SYSTEM_ERROR:** Issues with node configuration or bugs. The node logs details, updates metrics, and might attempt recovery or halt.

> - **SIGNATURE_INVALID:** If verification fails, SIGNATURE_INVALID is used. The block is marked invalid, and the BlockStatusManager triggers error-handling routines (requesting re-sends, removing corrupted data, notifying subscribers, etc.).

Reviewer: These extra newlines aren't needed and make the raw text a little harder to follow.

Author: ✅ Done!

### Signature invalid

> If the computed hash does not match the hash implied by the signature, the block is considered tampered. It is marked invalid and appropriate recovery steps are taken.

Reviewer: ```suggestion
If the computed hash does not match the hash signed by the network, the block is considered unverified. It is marked as such and publishers are requested to resend the block.
```

Author: Thank you!

✅ Done!

## Notes:
In order to keep it simple the first implementation will not start work async as items come, but will wait for the block_proof to arrive in order to start work async, that way we avoid a thread waiting for items and idle, and avoid the complexity of activating the async work as items come. The tradeoff of this simplification is that we might experience a higher latency for the stream of verified blocks, but we can always improve this in the future if is needed since the design already contemplates the async work of block items as they are received.

> By using metrics we will be able to measure the latency of block verification and decide if we need to improve it by activating the async work of block items as they come.

Reviewer: I would encourage you to start the async work up front for the following reasons:
1. We don't want to retain the whole block in memory, the hash algorithm laid out by Derek only keeps (for each of the two trees) the most recent item and up to 3 hash values, which is _much_ smaller (3-4 orders of magnitude).
1. If you use `CompletableFuture`, an input queue, a non-default fork-join work pool, and enable virtual threads it is quite easy to add asynchronous computation and the JVM will work-steal, which avoids any idle thread issues.
1. Any latency in verification is magnified and also creates risk because we cannot store a block until it's verified, and the publishers are waiting for acknowledgement that the block is verified before sending the next one.

Author: I actually agree with you on this one, but did the change at the request of @mattp-swirldslabs once we synced on the design after I share it on slack. but instead of changing the design, added notes to leave it as optional and the intention to optimize it at the future should the need arise. I think we should remove any latency where we can, but understand Matts view to want to keep things simple and safe when it comes to threads.

Author: I have modified the design doc so we keep async processing of hashes. That said, I will move this comments to a separate [discussion](https://github.com/hashgraph/hedera-block-node/discussions/400) on Module/Services System Loader and communication interface.

# Block Verification Design

## Table of Contents

1. [Purpose](#purpose)
1. [Goals](#goals)
1. [Terms](#terms)
1. [Entities](#entities)
1. [Design](#design)
1. [Interfaces](#interfaces)
1. [Enums](#enums)
1. [Metrics](#metrics)
1. [Exceptions](#exceptions)

## Purpose

The purpose of the Block Verification feature is to ensure that blocks received
from consensus nodes are valid and have not been tampered with. This is achieved
by re-calculating the block hash and verifying it against the signature provided
by the consensus node.

## Goals

1. The block-node must re-create the block hash from the block items and verify
   that it matches the hash implied by the signature.
1. If verification fails, the block should be considered invalid, and
   appropriate error-handling procedures must be triggered.

## Terms

<dl>
<dt>Consensus Node (CN)</dt><dd>A node that produces and provides blocks.</dd>
<dt>Block Items</dt><dd>The block data pieces (header, events, transactions, 
transaction result, state changes, proof) that make up a block.</dd>
<dt>Block Hash</dt><dd>A cryptographic hash representing the block’s integrity.</dd>

> <dt>Signature</dt><dd>The cryptographic signature of the block hash created by 
> Network private key (aka: LedgerId).</dd>
> <dt>Public Key</dt><dd>The public key of the LedgerId that signed the block.</dd>

Reviewer: The ledger ID is the public key (not private, which nobody actually knows) of the network, and all of the keys, effectively, identify the network.
```suggestion
<dt>Signature</dt><dd>The cryptographic signature on the block hash, created by the 
Network aggregation of "share" private keys, as described in the TSS design.</dd>
<dt>Public Key</dt><dd>The public key (a.k.a. Ledger ID) of the network that signed the block.</dd>
```

</dl>

## Entities

> - #### VerificationHandler

Reviewer 2: Wrong heading level for verification handler

```suggestion
- ### VerificationHandler
```

Author: Thanks! 

✅ Done!

    - Receives the stream of block items from the (partially) parsed and

>       unverified block items ring buffer.

Reviewer: The partially parsed character of items is, to my mind, an implementation detail, so perhaps leave that out?

Author: Removed it ✅
